---
title: "de Jong et al (2018)"
output:
  html_document:
    toc: yes
    toc_float: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# QTL mapping in MAGIC lines

This document contains the code to carry out QTL mapping in the MAGIC lines. 

We use the `happy.hbrem` package available from http://mtweb.cs.ucl.ac.uk/mus/www/magic/
and a custom R package, which is available from https://github.com/tavareshugo/MagicHelpR.

Details on how to use our package are available on its 
[documentation](https://tavareshugo.github.io/MagicHelpR/articles/introduction.html).

We first illustrate the basic QTL model fitting routine we use.

```{r, message = FALSE}
library(tidyverse)
library(patchwork)
library(MagicHelpR) # install with devtools::install_github("tavareshugo/MagicHelpR")

# Change ggplot2 default aesthetics
theme_set(theme_classic() + 
            theme(panel.grid = element_blank(), 
                  text = element_text(size = 8),
                  plot.title = element_text(hjust = -0.07, size = 10, face = "bold", 
                                            margin = margin(t = -5, b = 1)),
                  plot.subtitle = element_text(size = 10, hjust = 0.5)))
```


## Read data

### Phenotype data

We first read the phenotype data used for QTL mapping. This is using a table 
with all traits summarised per line (average trait values and plasticity).

```{r, message = FALSE}
# Read CSV file
magic_plas <- read_csv("../../data/phenotypes/magic_plasticity.csv")

# Keep only data for main traits of interest
magic_plas <- magic_plas %>% 
  select(id_kover, matches("height|bolt|totalbr"), -matches("rdpi"))
```

These are the trait distributions for all lines. We highlight those that 
flower early (< 25 days on LN), because those are the ones we focus on for 
the QTL mapping:

```{r}
magic_plas %>% 
  mutate(early_flower = bolt_mean_ln < 25) %>% 
  gather("trait", "value", -id_kover, -early_flower) %>% 
  group_by(trait) %>% 
  mutate(value_standard = (value - mean(value))/sd(value)) %>% 
  ggplot(aes(value_standard, fill = early_flower)) +
  geom_histogram() +
  facet_wrap(~ trait, scales = "free") +
  labs(x = "Standard Score", y = "Count", fill = "Early flowering\n(<25 days LN)")
```

From the Q-Q plots, we can see all traits fit well within a normal distribution, 
except flowering time when using all lines:

```{r}
# Q-Q plots for all lines
p1 <- magic_plas %>% 
  gather("trait", "value", -id_kover) %>% 
  group_by(trait) %>% 
  mutate(value_standard = (value - mean(value))/sd(value)) %>% 
  ggplot(aes(sample = value_standard)) +
  geom_qq() +
  facet_wrap(~ trait) +
  geom_abline() +
  labs(x = "Theoretical", y = "Observed", title = "All lines")

# Q-Q plots for early-flowering
p2 <- magic_plas %>% 
  filter(bolt_mean_ln < 25) %>% 
  gather("trait", "value", -id_kover) %>% 
  group_by(trait) %>% 
  mutate(value_mean = mean(value), value_standard = (value - mean(value))/sd(value)) %>% 
  ggplot(aes(sample = value_standard)) +
  geom_qq() +
  facet_wrap(~ trait) +
  geom_abline() +
  labs(x = "Theoretical", y = "Observed", title = "Early-flowering")

p1 + p2
```

We continue with the early-flowering lines only:

```{r}
magic_plas <- magic_plas %>% filter(bolt_mean_ln < 25)
nrow(magic_plas)
```


### Genotype data

The genotype data is available from http://mtweb.cs.ucl.ac.uk/mus/www/magic/

We use our own custom R package to download, tidy, and read these data for 
QTL mapping. Details of how to use these functions can be found in the 
package's introduction vignette (https://github.com/tavareshugo/MagicHelpR).

Here, we download the genotype data and read it into R:

```{r, message=FALSE, warning=FALSE}
# Download and tidy genotype data
if(!dir.exists("./temp_data")) dir.create("./temp_data")
downloadArabMagic(save_dir = "./temp_data/magic_genotypes/", tidy = TRUE)

# Read genotypes and add phenotypes
magic_gen <- magicFounderReconstruct("./temp_data/magic_genotypes/", 
                                     phenotypes = magic_plas, id = "id_kover")
```

We now have an object containg genotype and phenotype information for all our 
scored MAGIC lines:

```{r}
magic_gen
```

## QTL mapping

We exemplify the routine used for QTL mapping here. For the data presented 
in the paper, we used an script that was run non-interactively (this 
script is available upon request).

As an example, we use height, known to have a major QTL in the _Erecta_ gene.

```{r}
# Scan for QTL
height_ln <- scanQtl(magic_gen, "height_mean_ln")
```

This returns a `data.frame` object with the following results:

```{r, echo = FALSE}
height_ln %>% head %>% knitr::kable()
```

We can produce the usual manhattan plot like so:

```{r}
ggplot(height_ln, aes(bp/1e6, -log10(p))) +
  geom_line() +
  facet_grid(~ chromosome) +
  labs(x = "Mb") +
  geom_hline(yintercept = 3.5, linetype = "dashed")
```


## QTL figures

The above approach was used to do association mapping for all our traits of 
interest. Here, we read a CSV file containg all the results that were produced 
with our custom script.
These also include estimates of (computationally expensive) permutation-based 
p-values that control for the familywise error rate due to multiple-testing of 
~1K variants.

```{r}
qtl_results <- read_csv("../../data/qtl_magic/qtl_scans_prob.csv", guess_max = Inf)
head(qtl_results)
```

### Figure S5

QTL plot for height:

```{r}
#pdf("./figures/figureS5.pdf", width = 7, height = 3)
qtl_results %>% 
  filter(grepl("height", trait)) %>% 
  ggplot(aes(bp/1e6, -log10(p))) +
  geom_line() +
  facet_grid(trait ~ chromosome, scales = "free", space = "free_x") +
  labs(x = "Mb") +
  geom_hline(yintercept = 3.5, linetype = "dashed")
#dev.off()
```

This plot was further annotated manually in inkscape.

Notice that QTL on Chr5 for HN is due to a correlation between the two traits:

```{r}
cor.test(magic_plas$height_mean_hn, magic_plas$bolt_mean_hn, method = "pearson")
```

Here is just illustrating that if we fit a model taking flowering as a covariate 
the QTL on Chr5 disappears:

```{r}
height_hn <- scanQtl(magic_gen, phenotype = "height_mean_hn", covariates = "bolt_mean_hn")
qtl_results %>% 
  filter(grepl("height", trait)) %>% 
  ggplot(aes(bp/1e6, -log10(p))) +
  geom_line() +
  facet_grid(trait ~ chromosome, scales = "free", space = "free_x") +
  labs(x = "Mb") +
  geom_hline(yintercept = 3.5, linetype = "dashed")
```

### Figure 8

QTL plot for flowering:

```{r}
qtl_results %>% 
  filter(grepl("bolt", trait)) %>% 
  mutate(sig = ifelse(p_perm < 0.05, -log10(p), NA)) %>% 
  ggplot(aes(bp/1e6, -log10(p))) +
  geom_line() +
  geom_point(aes(y = sig), colour = "red3") +
  facet_grid(trait ~ chromosome, scales = "free", space = "free_x") +
  labs(x = "Mb") +
  geom_hline(yintercept = 3.5, linetype = "dashed")
```

QTL plot for branching:

```{r}
# Highlight peaks with permutation-based p-value < 0.05
qtl_results %>% 
  filter(grepl("totalbr", trait)) %>% 
  mutate(sig = ifelse(p_perm < 0.05, -log10(p), NA)) %>% 
  ggplot(aes(bp/1e6, -log10(p))) +
  geom_line(aes(linetype = covariate)) +
  geom_point(aes(y = sig), colour = "red3") +
  facet_grid(trait ~ chromosome, scales = "free", space = "free_x") +
  labs(x = "Mb") +
  geom_hline(yintercept = 3.5, linetype = "dashed") +
  scale_linetype_manual(values = c(3, 1))
```

Figure for paper (further annotated in inkscape):

```{r}
p1 <- qtl_results %>% 
  filter(grepl("bolt", trait)) %>% 
  separate(trait, c("trait", "stat", "nitrate"), sep = "_") %>% 
  mutate(nitrate = factor(nitrate, levels = c("hn", "ln", "D"))) %>% 
  ggplot(aes(bp/1e6, -log10(p))) +
  geom_line() +
  facet_grid(trait + nitrate ~ chromosome, scales = "free", space = "free") +
  labs(x = "Mb") +
  geom_hline(yintercept = 3.5, linetype = "dashed", colour = "grey") +
  theme(panel.grid = element_blank()) +
  scale_y_continuous(breaks = seq(0, 10, 2)) +
  scale_x_continuous(breaks = seq(0, 30, 10))

p2 <- qtl_results %>% 
  filter(grepl("totalbr", trait)) %>% 
  separate(trait, c("trait", "stat", "nitrate"), sep = "_") %>% 
  mutate(nitrate = factor(nitrate, levels = c("hn", "ln", "D"))) %>% 
  ggplot(aes(bp/1e6, -log10(p))) +
  geom_line(aes(linetype = covariate)) +
  facet_grid(trait + nitrate ~ chromosome, scales = "free", space = "free") +
  labs(x = "Mb") +
  geom_hline(yintercept = 3.5, linetype = "dashed", colour = "grey") +
  theme(panel.grid = element_blank()) +
  scale_linetype_manual(values = c(3, 1)) +
  theme(legend.position = "none") +
  scale_y_continuous(breaks = seq(0, 10, 2)) +
  scale_x_continuous(breaks = seq(0, 30, 10))

#pdf("./figures/figure8.pdf", width = 7, height = 6)
p1 + p2 + plot_layout(ncol = 1, heights = c(2/5, 3/5))
#dev.off()
```

Correlation between total branches and flowering time, to include in figure 
legend:

```{r}
magic_plas <- read_rds("../../data/phenotypes/magic_plasticity.rds") %>% 
  filter(bolt_mean_ln < 25)

# Correlation between branching and bolting
cor.test(magic_plas$totalbr_mean_D, magic_plas$bolt_mean_ln, method = "pearson")
cor.test(magic_plas$totalbr_mean_ln, magic_plas$bolt_mean_ln, method = "pearson")
cor.test(magic_plas$totalbr_mean_hn, magic_plas$bolt_mean_hn, method = "pearson")
```


## Estimating QTL effect sizes - Fig. 10

The QTL effect can be estimated by using the phenotype imputation method detailed 
in [Kover .. Mott (2009)](http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1000551).

For example, for the _Erecta_ peak for height shown above, this could be done 
like so:

```{r}
# Get peak marker name
hgt_ln_2 <- qtl_results %>% filter(trait == "height_mean_ln") %>% 
  arrange(p) %>% slice(1) %>% .$marker

# Make a plot, adding SNP alleles of each founder acccession
p1 <- estimateFounderEffect(magic_gen, "height_mean_ln", hgt_ln_2) %>% 
  ggplot(aes(accession, effect_mean)) +
  geom_hline(yintercept = 0, linetype = "dashed") +
  geom_pointrange(aes(ymax = effect_up, ymin = effect_lo)) +
  geom_text(data = getFounderGenotypes(magic_gen)[[hgt_ln_2]], 
            aes(y = -2.5, label = allele, colour = allele), size = 2.5) +
  theme_bw() +
  theme(text = element_text(size = 8), legend.position = "none", axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        panel.grid.major.x = element_blank()) +
  scale_y_continuous(breaks = -5:5, limits = c(-2.5, 2.5)) + 
  scale_colour_manual(values = c("black", "red3")) +
  labs(x = "Accession", y = "Standardized Effect", title = hgt_ln_2)
p1
```

And we can clearly see that the _Ler_ allele is associated with particularly 
small plants (~2 standard deviations below the overall average). 

We also annotated the plot with the SNP alleles of each founder 
accession. This reveals how the _Ct_ accession actually shares an allele with 
_Ler_, despite having a clearly different contribution to the phenotype. 

The rest of Fig. S8:

```{r}
ft_hn_5 <- qtl_results %>% 
  filter(trait == "bolt_mean_hn" & chromosome == 5) %>% 
  arrange(p) %>% slice(1) %>% .$marker

p2 <- estimateFounderEffect(magic_gen, "bolt_mean_hn", ft_hn_5) %>% 
  ggplot(aes(accession, effect_mean)) +
  geom_hline(yintercept = 0, linetype = "dashed") +
  geom_pointrange(aes(ymax = effect_up, ymin = effect_lo)) +
  geom_text(data = getFounderGenotypes(magic_gen)[[ft_hn_5]], 
            aes(y = -2.5, label = allele, colour = allele), size = 2.5) +
  theme_bw() +
  theme(text = element_text(size = 8), legend.position = "none", axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        panel.grid.major.x = element_blank()) +
  scale_y_continuous(breaks = -5:5, limits = c(-2.5, 2.5)) + 
  scale_colour_manual(values = c("red3", "black")) +
  labs(x = "Accession", y = "Standardized Effect", title = ft_hn_5)

p2
```

```{r}
totalbr_d_2 <- qtl_results %>% 
  filter(trait == "totalbr_mean_D" & chromosome == 2) %>% 
  arrange(p) %>% slice(1) %>% .$marker

p3 <- estimateFounderEffect(magic_gen, "totalbr_mean_D", totalbr_d_2) %>% 
  ggplot(aes(accession, effect_mean)) +
  geom_hline(yintercept = 0, linetype = "dashed") +
  geom_pointrange(aes(ymax = effect_up, ymin = effect_lo)) +
  geom_text(data = getFounderGenotypes(magic_gen)[[totalbr_d_2]], 
            aes(y = -2.5, label = allele, colour = allele), size = 2.5) +
  theme_bw() +
  theme(text = element_text(size = 8), legend.position = "none", axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        panel.grid.major.x = element_blank()) +
  scale_y_continuous(breaks = -5:5, limits = c(-2.5, 2.5)) + 
  scale_colour_manual(values = c("red3", "black")) +
  labs(x = "Accession", y = "Standardized Effect", title = totalbr_d_2)
p3
```


This shows the effect of the main QTL for height, shoot branching plasticity and 
flowering QTL on chr 5 presented in Fig 10:

```{r}
#pdf("./figures/figure10.pdf", height = 3, width = 7.5)
p1 + p2 + p3
#dev.off()
```

Also get the effect of the two QTL that were observed with added covariate for 
branching on LN:

```{r}
# Get peak marker name
totalbr_ln_cov <- qtl_results %>% 
  filter(trait == "totalbr_mean_ln" & covariate == "bolt age" & chromosome %in% c(1, 3)) %>% 
  group_by(chromosome) %>% 
  arrange(p) %>% slice(1) %>% .$marker

# Make a plot, adding SNP alleles of each founder acccession
p1 <- estimateFounderEffect(magic_gen, "totalbr_mean_ln", totalbr_ln_cov[1], 
                            standardised = FALSE, covariates = "bolt_mean_ln") %>% 
  ggplot(aes(accession, effect_mean)) +
  geom_hline(yintercept = 0, linetype = "dashed") +
  geom_pointrange(aes(ymax = effect_up, ymin = effect_lo)) +
  geom_text(data = getFounderGenotypes(magic_gen)[[totalbr_ln_cov[1]]], 
            aes(y = -1, label = allele, colour = allele), size = 2.5) +
  theme_bw() +
  theme(legend.position = "none", axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        panel.grid.major.x = element_blank()) +
  scale_colour_manual(values = c("black", "red3")) +
  labs(x = "Accession", y = "Residual Effect", title = totalbr_ln_cov[1]) +
  scale_y_continuous(breaks = seq(-1, 2, 1), limits = c(-1, 1.6))

# Make a plot, adding SNP alleles of each founder acccession
p2 <- estimateFounderEffect(magic_gen, "totalbr_mean_ln", totalbr_ln_cov[2], 
                            standardised = FALSE, covariates = "bolt_mean_ln") %>% 
  ggplot(aes(accession, effect_mean)) +
  geom_hline(yintercept = 0, linetype = "dashed") +
  geom_pointrange(aes(ymax = effect_up, ymin = effect_lo)) +
  #geom_label(aes(label = n)) +
  geom_text(data = getFounderGenotypes(magic_gen)[[totalbr_ln_cov[2]]], 
            aes(y = -1, label = allele, colour = allele), size = 2.5) +
  theme_bw() +
  theme(legend.position = "none", axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        panel.grid.major.x = element_blank()) +
  scale_colour_manual(values = c("black", "red3")) +
  labs(x = "Accession", y = "Residual Effect", title = totalbr_ln_cov[2]) +
  scale_y_continuous(breaks = seq(-1, 2, 1), limits = c(-1, 1.6))

p1 + p2
```


## Extra: comparing haplotype and SNP-based method

The QTL table of results also contains results of fitting the model using SNP
genotypes. This can be achieved with the `magicQtlScan()` function, using the 
option `method = "allele"`. For example, for height this could be done like so:

```{r}
# Make the QTL scan
height_ln_snp <- scanQtl(magic_gen, "height_mean_ln", method = "allele")

# Produce the plot
ggplot(height_ln_snp, aes(bp/1e6, -log10(p))) +
  geom_point() +
  facet_grid(~ chromosome, scales = "free_x", space = "free_x") +
  labs(x = "Mb") +
  geom_hline(yintercept = 3.5, linetype = "dashed")
```

Since the table already contains the p-values for the test using SNP genotypes, 
we can plot them overlapping the probability-based genotypes. 

The example below demonstrates how the haplotype-based method of mapping gives 
peaks that do not always coincide with SNP-based method:

```{r}
# Manhattan plot with two QTL fitting routines
p1 <- qtl_results %>% 
  filter(trait == "totalbr_mean_D" & covariate == "none") %>% 
  mutate(sig = ifelse(p_perm_snp < 0.05, -log10(p_snp), NA)) %>% 
  ggplot(aes(bp/1e6, -log10(p))) +
  geom_point(aes(y = -log10(p_snp)), colour = "grey") +
  geom_point(aes(y = sig)) +
  #geom_point(data = qtl_snp2, aes(y = sig), colour = "red3") +
  geom_line() +
  facet_grid( ~ chromosome, scales = "free", space = "free_x") +
  labs(x = "Mb") +
  geom_hline(yintercept = 3.5, linetype = "dashed") +
  theme(panel.grid = element_blank())

# Zoomed-in version of previous plot
p2 <- qtl_results %>% 
  filter(trait == "totalbr_mean_D" & chromosome == 2 & covariate == "none") %>% 
  mutate(sig = ifelse(p_perm_snp < 0.05, -log10(p_snp), NA)) %>% 
  ggplot(aes(bp/1e6, -log10(p))) +
  #geom_point(data = qtl_snp2, aes(y = sig), colour = "red3") +
  geom_line() +
  geom_point(aes(y = -log10(p_snp)), colour = "grey") +
  geom_point(aes(y = sig)) +
  labs(x = "Mb") +
  geom_hline(yintercept = 3.5, linetype = "dashed") +
  theme(panel.grid = element_blank()) +
  xlim(17.5, 20)

# Example of SNP effect
p3 <- estimateFounderEffect(magic_gen, "totalbr_mean_D", "MAX3_141") %>% 
  ggplot(aes(accession, effect_mean)) +
  geom_hline(yintercept = 0, linetype = "dashed") +
  geom_pointrange(aes(ymax = effect_up, ymin = effect_lo), size = 0.2) +
  #geom_label(aes(label = n)) +
  geom_text(data = getFounderGenotypes(magic_gen)[["MAX3_141"]], 
            aes(y = -1.2, label = allele, colour = allele), size = 2) +
  theme(legend.position = "none", axis.text.x = element_text(angle = 90, hjust = 1), panel.grid.major.x = element_blank()) +
  scale_colour_manual(values = c("red3", "black")) +
  labs(x = "Accession", y = "Standardized Effect", title = "MAX3_141")

# Another example of SNP effect
p4 <- estimateFounderEffect(magic_gen, "totalbr_mean_D", "SOC1_651") %>% 
  ggplot(aes(accession, effect_mean)) +
  geom_hline(yintercept = 0, linetype = "dashed") +
  geom_pointrange(aes(ymax = effect_up, ymin = effect_lo), size = 0.2) +
  #geom_label(aes(label = n)) +
  geom_text(data = getFounderGenotypes(magic_gen)[["SOC1_651"]], 
            aes(y = -1.2, label = allele, colour = allele), size = 2) +
  theme(legend.position = "none", axis.text.x = element_text(angle = 90, hjust = 1), panel.grid.major.x = element_blank()) +
  scale_colour_manual(values = c("red3", "black")) +
  labs(x = "Accession", y = "Standardized Effect", title = "SOC1_651")


p1 /
{p3 + p2 + p4}
  
#grid.arrange(p1, arrangeGrob(p3, p2, p4, ncol = 3), p5, ncol = 1)
```

For the peak on Chr2 (highlighted in the bottom middle panel), there is one 
significant SNP (in black), whose effects are shown on the left. Its estimated 
effect sizes are plotted on the left panel. We can see that in this case the two 
accessions with the most extreme effects (Bur and Sf) both share the same SNP 
allele - thus becoming "signicant". However, there is a substantial range of effects 
that these two SNP alleles do not capture (from Ler to Tsu, there is substantial 
difference). 

This illustrates well why there is a case for multiple alleles contributing to this 
trait, as we argue in the paper.


# Session Info

```{r}
sessionInfo()
```

