---
title: "de Jong et al (2018)"
output:
  html_document:
    toc: yes
    toc_float: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# QTL mapping in accessions

This document contains the code to produce manhattan plots of the GWAS run on 
accessions and heritability estimates from SNPs.

These analysis were carried using [GCTA](http://cnsgenomics.com/software/gcta/).
Shell scripts for these analysis are provided separately.

```{r}
library(tidyverse)
library(patchwork)

# Change ggplot2 default aesthetics
theme_set(theme_classic() + 
            theme(panel.grid = element_blank(), 
                  text = element_text(size = 8),
                  plot.title = element_text(hjust = -0.07, size = 10, face = "bold", 
                                            margin = margin(t = -5, b = 1)),
                  plot.subtitle = element_text(size = 10, hjust = 0.5)))

```


## Fig S6 - GWA heritability

```{r}
her <- read_csv("../../data/gwas_accessions/compiled_heritabilities.csv")
```

This is the heribitability estimated from genetic relatedness:

```{r}
#pdf("./figures/figureS6.pdf", width = 4.5, height = 3)
her %>% 
  filter(trait %in% c("bolt", "height", "totalbr", "totalsil")) %>% 
  mutate(trait = ifelse(stage == "sen", paste0(trait, "\n(sen)"), trait),
         nitrate = ifelse(nitrate == "D", "Plasticity", toupper(nitrate)),
         err_up = ifelse(her_var + her_se < 1, her_var + her_se, 1),
         err_lo = ifelse(her_var - her_se > 0, her_var - her_se, 0)) %>% 
  ggplot(aes(trait, her_var, fill = nitrate)) +
  geom_bar(stat = "identity", position = position_dodge(0.9), aes(colour = nitrate)) +
  #geom_point(aes(colour = nitrate), position = position_dodge(0.9)) +
  geom_errorbar(aes(ymax = err_up, ymin = err_lo), width = 0, position = position_dodge(width = 0.9)) +
  facet_grid(set ~ .) +
  scale_fill_manual(values = c("grey48", "grey", "black")) +
  scale_colour_manual(values = c("grey48", "grey", "black")) +
  theme(panel.border = element_rect(colour = "black", fill=NA)) +
  labs(fill = "Nitrate")
#dev.off()
```

Here's a view of the raw values (instead of relative fractions):

```{r}
her %>% 
  filter(trait %in% c("totalbr")) %>% 
  mutate(trait = ifelse(stage == "sen", paste0(trait, "\n(sen)"), trait),
         nitrate = ifelse(nitrate == "D", "Plasticity", toupper(nitrate)),
         err_up = ifelse(her_var + her_se < 1, her_var + her_se, 1),
         err_lo = ifelse(her_var - her_se > 0, her_var - her_se, 0)) %>% 
  select(trait, nitrate, genetic_var, residual_var) %>% 
  gather("var", "value", genetic_var, residual_var) %>% 
  ggplot(aes(trait, value, fill = var)) +
  geom_bar(stat = "identity") +
  facet_grid(. ~ nitrate) +
  scale_fill_manual(values = c("grey48", "grey", "black")) +
  labs(fill = "Variance component", x = "", y = "Variance")
```


## Fig 9 - Manhattan plots

```{r}
# Read GWAS results
## produced with GCTA software with LOCO method for relatedness correction
gwas <- read_csv("../../data/gwas_accessions/compiled_gwas.csv")

# Set genome-wide threshold using bonferroni correction
## The number of SNPs from the 250K snp set is used
thr <- gwas %>% filter(trait == "bolt" & nitrate == "hn" & set == "snp250k") %>% nrow
thr <- 0.05/thr
```

This gives a threshold of `r -log10(thr)`.

This is the manhattan plot shown in Fig. 9 of the paper:

```{r}
#pdf("./figures/figure9.pdf", width = 7, height = 6)
gwas %>% 
  filter(trait %in% c("bolt", "totalbr") & set == "snp250k" & p < 0.01 & stage == "silique") %>% 
  mutate(nitrate = ifelse(nitrate == "D", "Plasticity", toupper(nitrate))) %>% 
  ggplot(aes(bp/1e6, -log10(p))) +
  geom_point(size = 0.5) +
  facet_grid(trait + nitrate ~ Chr, scales = "free_x", space = "free_x") +
  geom_hline(yintercept = -log10(thr), linetype = "dashed", colour = "grey") +
  theme(panel.grid = element_blank()) +
  scale_x_continuous(breaks = seq(0, 30, 10)) +
  labs(x = "Position (Mb)")
#dev.off()
```

Just for comparison, here's the result with the imputed SNPs, but notice that 
this would require a different genome-wide threshold (showing the 250k threshold):

```{r}
gwas %>% 
  filter(trait %in% c("bolt", "totalbr") & set == "imputed" & p < 0.01 & stage == "silique") %>% 
  mutate(nitrate = ifelse(nitrate == "D", "Plasticity", toupper(nitrate))) %>% 
  ggplot(aes(bp/1e6, -log10(p))) +
  geom_point(size = 0.5) +
  facet_grid(trait + nitrate ~ Chr, scales = "free_x", space = "free_x") +
  geom_hline(yintercept = -log10(thr), linetype = "dashed", colour = "grey") +
  theme(panel.grid = element_blank()) +
  scale_x_continuous(breaks = seq(0, 30, 10)) +
  labs(x = "Position (Mb)")
```

Also, the GWAS for senescence traits (nothing significant):

```{r}
gwas %>% 
  filter(trait %in% c("totalsil", "totalbr") & set == "snp250k" & p < 0.01 & stage == "senescence") %>% 
  mutate(nitrate = ifelse(nitrate == "D", "Plasticity", toupper(nitrate))) %>% 
  ggplot(aes(bp/1e6, -log10(p))) +
  geom_point(size = 0.5) +
  facet_grid(trait + nitrate ~ Chr, scales = "free_x", space = "free_x") +
  geom_hline(yintercept = -log10(thr), linetype = "dashed", colour = "grey") +
  theme(panel.grid = element_blank()) +
  scale_x_continuous(breaks = seq(0, 30, 10)) +
  labs(x = "Position (Mb)")
```

This graph shows the Q-Q plots for every GWAS, together with the manhattan plot 
for the only significant case:

```{r}
# Q-Q plots
p1 <- gwas %>% 
  filter(trait %in% c("totalsil", "totalbr", "height", "bolt") & set == "snp250k") %>% 
  mutate(nitrate = ifelse(nitrate == "D", "Plasticity", toupper(nitrate)),
         trait = ifelse(stage == "senescence", paste0(trait, " (sen)"), trait)) %>% 
  group_by(trait, nitrate) %>% 
  mutate(pexp = -log10( ppoints(n()) ),
         pobs = -log10(sort(p, decreasing = FALSE)),
         sig = ifelse(pobs > -log10(thr), pobs, NA)) %>% 
  ggplot(aes(pexp, pobs)) + 
  geom_point(colour = "grey") +
  geom_point(aes(y = sig), colour = "red3") +
  geom_abline() +
  geom_hline(yintercept = -log10(thr), linetype = "dashed", colour = "grey") +
  facet_grid(trait ~ nitrate) +
  labs(x = expression(Expected~~-log[10](italic(p))),
       y = expression(Observed~~-log[10](italic(p))))

# GWAS for significant SNP
p2 <- gwas %>% 
  filter(trait == "bolt" & nitrate == "hn" & set == "snp250k" & p < 0.01 & stage == "silique") %>% 
  mutate(sig = ifelse(p < thr, "sig", "non-sig")) %>% 
  ggplot(aes(bp/1e6, -log10(p), colour = sig)) +
  geom_point(size = 0.5) +
  facet_grid( ~ Chr, scales = "free_x", space = "free_x") +
  geom_hline(yintercept = -log10(thr), linetype = "dashed", colour = "grey") +
  theme(panel.grid = element_blank()) +
  scale_x_continuous(breaks = seq(0, 30, 10)) +
  labs(x = "Position (Mb)") +
  scale_colour_manual(values = c("black", "red3")) +
  theme(legend.position = "none")

# Plot them together
p1 + p2 + plot_layout(ncol = 1, heights = c(3/4, 1/4))
```


## Extra: Gene-level GWAs

We also ran a gene-based analysis, using several SNPs centered around genes 
(+/- 5Kb around each gene annotation). This uses the set-based method implemented 
in GCTA from [Bakshi .. Yang (2016)](https://www.nature.com/articles/srep32894).

```{r}
# Read GWAS results
## produced with GCTA software with LOCO method for relatedness correction
gwas_gene <- read_csv("../../data/gwas_accessions/compiled_gwas_gene_level.csv")

# Set genome-wide threshold using bonferroni correction
## The number of SNPs from the 250K snp set is used
thr_gene <- gwas_gene %>% filter(trait == "bolt" & nitrate == "hn" & set == "snp250k") %>% nrow
thr_gene <- 0.05/thr_gene
```

This is the distribution of number of SNPs per gene (more SNPs in the imputed 
set, of course):

```{r}
gwas_gene %>% 
  filter(trait %in% c("bolt", "totalbr")) %>% 
  ggplot(aes(No.SNPs)) +
  geom_density() +
  facet_grid(trait + nitrate ~ set, scales = "free_x", space = "free_x")
```

No significant hits again:

```{r}
gwas_gene %>% 
  filter(trait %in% c("bolt", "totalbr") & set == "snp250k" & Pvalue < 0.01) %>% 
  mutate(nitrate = ifelse(nitrate == "D", "Plasticity", toupper(nitrate))) %>% 
  ggplot(aes(Start/1e6, -log10(Pvalue))) +
  geom_point(size = 0.5) +
  facet_grid(trait + nitrate ~ Chr, scales = "free_x", space = "free_x") +
  geom_hline(yintercept = -log10(thr_gene), linetype = "dashed", colour = "grey") +
  theme(panel.grid = element_blank()) +
  scale_x_continuous(breaks = seq(0, 30, 10)) +
  labs(x = "Position (Mb)")
```

Even using the imputed set of SNPs (more SNPs per gene):

```{r}
gwas_gene %>% 
  filter(trait %in% c("bolt", "totalbr") & set == "imputed" & Pvalue < 0.01) %>% 
  mutate(nitrate = ifelse(nitrate == "D", "Plasticity", toupper(nitrate))) %>% 
  ggplot(aes(Start/1e6, -log10(Pvalue))) +
  geom_point(size = 0.5) +
  facet_grid(trait + nitrate ~ Chr, scales = "free_x", space = "free_x") +
  geom_hline(yintercept = -log10(thr_gene), linetype = "dashed", colour = "grey") +
  theme(panel.grid = element_blank()) +
  scale_x_continuous(breaks = seq(0, 30, 10)) +
  labs(x = "Position (Mb)")
```

# Session info

```{r}
sessionInfo()
```

